### PR TITLE
Add options to debug collisions with gdscript-godot-run-project-debug

### PR DESCRIPTION
**Pull request by VlachJosef**
Running `gdscript-godot-run-project-debug` with prefix argument allows to pass extra command line options to godot executable for debugging `collision/navigation`, namely:

  1) `<no options>` _(default)_
  2) `--debug-collisions`
  3) `--debug-navigation`
  4) `--debug-collisions --debug-navigation`

### CODE REVIEW

**Comment by NathanLovato** [comment]
Thanks for the contribution, but I don't think it's a good way to launch the project because it's neither intuitive nor discoverable to have to do things like:

> `C-u 1 C-c C-r C-d`

I would rather recommend users to use one of Emacs's shells with autocompletion (there's shell autocompletion for bash and fish for Godot).

I'm also not sure running the game with debugging options but without the editor and without the debugger running is going to be a common use case, thus one we should support.

So I'm not sold on the PR for now.

**Comment by VlachJosef** [comment]
You don't need to specify prefix argument value to `1` as in:

> C-u 1 C-c C-r C-d

Simple use more ergonomic:

> C-u C-c C-r C-d

Prefix argument value is `4` when not specified explicitly, but its value doesn't matter at the moment, only that it exists.

Explicit value could be utilise to pick debug option directly, though, without need to open selection in mini-buffer.

This PR is not to replace shell with autocompletion at all. All it aims at are users who are looking for checkboxes in Godot editor under Debug menu, they are:

```
 - [ ] Visible collisions shapes
 - [ ] Visible Navigation
 ```

without need to leave emacs.

> I'm also not sure running the game with debugging options but without the editor and without the debugger running is going to be a common use case, thus one we should support.

When this would be a case, why `gdscript-godot-run-project-debug` (`C-c C-r C-d`) exists in the first place? 

I'm new to Godot and I didn't need to use debugger so far, but I'd already appreciate to have a quick way to visualise collisions.

**Comment by NathanLovato** [comment]
The part I don't like is the use of the prefix argument, that is to say, requiring the user to learn a number by heart to launch a scene in debug mode with a given option. Sorry if I didn't make that clear.

I would just like a design that is more discoverable than that. As it stands, it's neither ergonomic nor intuitive (and none of our shortcuts are ergonomic, that's something we need to work on).

The comparison I made with the shell is that on the shell, you wouldn't have to know `C-u` or `C-u 2`. Instead, you would type `--d`, press tab, and get a list of debug options.

Screenshot from 2020-05-27 07-32-49

This is discoverable and intuitive because no option is hidden from me.

The thing is, I don't know the Emacs-y way of providing such a feature. There are right-click menus but you can deactivate them. I personally use swiper/ivy but it's not there by default.

Suggestions are welcome! And if there's no alternative, I guess the default argument could be better than nothing.

@ofrank123 would you have any idea on the matter?

**Comment by VlachJosef** [comment]
I agree that use of prefix argument (ie. abuse of mini-buffer to simulate checkboxes) is not standard and it is not intuitive.

Can I propose to use Hydra as an replacement for that? It would solve problem with ergonomic shortcuts as well. Downside is that it would be extra package as a dependency, but again, no one would be forced to use it unless they want that extra features and better ergonomy.

I think it could look similar to this. There may be two hydras (ie. tabs):

First hydra could look like this:
```
A Run options *       B Debug options
--------------------------------------------------------------
d [ ] Debug     p run current project    c run current script
e [ ] Editor    s run current scene      

```

`A`, `B`, `d`, `e`, `p`, `s` and `c` are hotkeys on this hydra. 

`d` and `e` would behave as checkboxes, so it would be rendered either like

```
d [ ] Debug
```
if not selected or 
```
d [X] Debug
```
if selected.

This hydra should cover all commands from gdscript-godot.el ie.:

```
gdscript-godot-open-project-in-editor
gdscript-godot-run-*
gdscript-godot-edit-*
```

Second hydra (accessed by pressing `B` key) could be aditional debug options to pass to godot executable (if they would make a sense):

```
A Run options        B Debug options *
--------------------------------------------------------------
 c [ ] Visible collisions shapes
 n [ ] Visible navigation
```

What do you think? Would you be interested in this approach?

Hydra could be useful also for other tasks. For example I'd like to see hydra with option to quickly switch to `*godot*` buffer:

```
A Run options *       B Debug options
--------------------------------------------------------------
d [ ] Debug     p run current project    c run current script
e [ ] Editor    s run current scene      g switch to *godot*
```

**Comment by NathanLovato** [comment]
I'd love to use Hydra, but I think we can't have it in here: you can't have external dependencies on MELPA programs, you need to create separate packages. For example, we need a separate flycheck-gdscript package for flycheck integration, or probably yas-gdscript for yasnippet. That's also why even though I use evil, yas, and which-key, I didn't put the related configuration in the package.

**Comment by VlachJosef** [comment]
There are packages which depends on Hydra on MELPA. For example [pretty-hydra.el](https://github.com/jerrypnz/major-mode-hydra.el/blob/master/pretty-hydra.el#L8).
I think people create separate integration packages as a way to tackle complexity, so that not everything is in one package etc... It is not because of MELPA's restrictions.

There is also no need to declare Hydra as a dependency at all. We can handle missing Hydra gracefully in source .el file itself. For example by declaring

```
(require 'hydra "hydra" 't) ;; fail silently
```

or like this: 

```
(unless (fboundp 'defhydra)
    (error "This function requires Hydra package to be installed."))
```

And then it is only a matter of explaining it in the documentation (error message) that this feature needs extra package.

Having said all of that there is also [magit-popup](https://magit.vc/manual/magit-popup.html) package which we can consider to use instead of Hydra. In any case question of integration remains. But since gdscript-mode is still fairly small I'd say it is ok to have this feature added into it without burden which would be caused by having to maintain second package.

**Comment by NathanLovato** [comment]
>  I think people create separate integration packages as a way to tackle complexity, so that not everything is in one package etc... It is not because of MELPA's restrictions.

Okay, that's good to know. Somehow I thought they asked people to split add-ons to other packages into separate packages. I remember reading something like that, about splitting your package if it had dependencies, in the contributing guide, but I might've misunderstood the meaning.

With that graceful dependency check, I'm all for having that config! Maybe more with hydra rather than magit, as a user who doesn't work with git may not have it? And because e.g. Doom emacs and spacemacs use different tools for popups or windows like these.

**Comment by VlachJosef** [comment]
I've found that part you've mentioned. It says:

> - MELPA *only* looks at the main `.el` file (or `-pkg.el` file, if
  provided, though we discourage that).  If you have different
  optional parts of a large package, each of which has different
  dependencies, then these should really be published as *separate
  packages*.

So yes, they are discouraging what I'm proposing, but on the other hand `gdscript-completion.el` already does something similar with a dependency on a `projectile` (albeit there is a fallback in this case).

Bottom line is everything will work, people only won't see `projectile` or `hydra` listed as a dependency on a Melpa.

So if this didn't change your mind I'll be glad to crack on with that Hydra integration?

**Comment by NathanLovato** [comment]
> So if this didn't change your mind I'll be glad to crack on with that Hydra integration?

Go ahead! I think it's much easier to manage all that functionality in one package. It's easier for people to find it and contribute.

File: gdscript-godot.el [hunk 1]
```diff
@@ -36,6 +36,14 @@
 (require 'gdscript-utils)
 
 ;;;###autoload
+(defvar gdscript-godot-debug-selected-option 1)
```
Comment (NathanLovato) [inline gdscript-godot.el:39]: There's a small thing I'd need you to change in your PRs: anything that's pseudo-private (shouldn't be accessed by the user) should be with two dashes `package-name--symbol-name`. It's part of the MELPA style guide. Could you change the variables and functions you added to be `gdscript-godot--*`?

**Comment by NathanLovato** [comment]
Just tested the PR, it's working like a charm! Thanks. I just left a comment about naming conventions for MELPA.

**Comment by NathanLovato** [comment]
I'll merge now and rename the few functions. Thanks much for your contributions once again!
